### DISCUSSION TITLE
Valetudo 2024.06.2

### DISCUSSION THREAD

**Discussion by Hypfer** [Hypfer/Valetudo#2078] [Release Discussions]
<div align="center">
    <img src="https://raw.githubusercontent.com/Hypfer/Valetudo/master/assets/logo/valetudo_logo_with_name.svg" width="800" alt="valetudo">
    <p align="center"><h2>2024.06.2</h2></p>
</div>

Exciting new robots, lots of UI improvements, new features and some words on not burning out

## Valetudo 2024.06.2

This version bumps the MQTT.js library once again to hopefully fix some more keepalive-related issues that apparently were caused by the prior fixes. They were causing issues in _different_ setups than previously though, which I'd call a win 👍

Nothing else has changed, so if you've already read the 2024.06.1 changelog, nothing new to see here.
If not, feel free to continue reading.

## Valetudo 2024.06.1

Since I did not personally buy an X40 for this release, I kinda expected that there would be some minor additional issues that would show up as soon as someone attempts to root one. And indeed, there was one.

Specifically, we found out that the EU variant(?) is assigned a different Model ID than the US one.
Fortunately though, everything else seems to be the same as on the US model (which was the one we've tested prior to the release).

After then employing highly advanced OSINT research skills (thanks, UHD YouTube reviews), I've discovered that there are two additional Model ID identities the X40 might have.

This release is just 2024.06.0 plus three strings, enabling Valetudo to start on these robots.
Thus, if you have already read the 2024.06.0 release notes, you can skip the rest of this.

If not, feel free to continue reading.

## New Robots

Since there has been quite some innovation going on in the robot space since the last time we've released support for a new robot, we figured that it would be time to take a closer look at new robots once more.

Unfortunately, with innovation also came a steep increase in price. But, after irresponsibly spending multiple thousand euros on new robots to void their warranty 30s after they've left the box, I have good news:

With the Dreame X40 and L10s Pro Ultra heat now being supported, we have Valetudo support for what according to reviewers and also our own experience is the best vacuum robot money can buy you right now.

![image](https://github.com/Hypfer/Valetudo/assets/974410/51d607d8-ae9f-4207-8a03-a45ce320bdbd)

![image](https://github.com/Hypfer/Valetudo/assets/974410/4a6507b4-2caa-4c6c-87a9-2d959f2d10f7)

This sounds amazing and also is amazing but do keep in mind that the previously recommended L10s Ultra already is a very good vacuum robot. There aren't many things to improve about it, and, frankly, there are no improvements that would warrant spending more than twice the money.

Still, there have been _notable_ improvements:
- Mops that extend to the side beyond the robot for better edge cleaning
- X40 Only: A side brush that extends beyond the robot for better corner and edge cleaning
- X40 Only: An integrated LED flashlight for much improved AI recognition in all lighting conditions
- X40 Only: Mops that can be automatically detached and left in the dock
- An improved dock design featuring bigger water tanks and easier cleaning
- The ability to auto-empty multiple times during a cleanup
- General cleaning/navigation logic updates
- Updated AI models

As always, don't forget that Valetudo is Valetudo.
It does not strive to implement each and every feature the vendor app might offer. In fact, it strives to explicitly not do that.

## UI improvements

Apart from support for these robots, a lot of things have changed about Valetudo itself too.

### Custom Icons

Valetudo now has a custom icon set for all presets (Fan, Water, Operating mode).

![image](https://github.com/Hypfer/Valetudo/assets/974410/532c2591-03d4-4368-914e-f6c3a4af0fa4)

Thanks a lot to everyone who took part in the iterative feedback process that led to these! :)
I think they really bring the UI together nicely. And, as an added bonus, they also work better on narrow screens.

### Controls polish

Speaking of bringing the UI together nicely, the controls have all received some love.

![image](https://github.com/Hypfer/Valetudo/assets/974410/d3325d74-4746-463f-893b-2126745ce6b5)

- They're now more unified, using the same padding and spacing and elements.
- They now better show feedback to the user (check the dock controls or presets with a bad wifi connection)
- And, the state flag display now doesn't clip anymore on very narrow screens.

### HiDPI support

Screens bring us to the next topic:
Did you ever notice that on mobile, the map view looks a bit bad, blurry and low res?

You might've thought that it's for performance reasons, but it's actually due to this new and amazing innovation called HiDPI screens that has just been released… almost 15 years ago.
Starting with this release, the Valetudo Map renderer now also takes advantage of this revolutionary display technology.

![image](https://github.com/Hypfer/Valetudo/assets/974410/584ee6c6-e899-4702-ad0f-28dc78b5bb66)

Do note though that unfortunately, HiDPI support had to be disabled for Safari due to Safari being Safari.

### Mobile map font rendering

Apart from HiDPI support, another thing that has been fixed in the map renderer is font rendering on firefox mobile.

![image](https://github.com/Hypfer/Valetudo/assets/974410/e61d6289-6d3d-4615-8be2-6ca90214cb6e)

Previously, fill and stroke were called in the wrong order, leading to those overlaps.
You can also see that this screenshot was taken before HiDPI support was implemented, because the text is still somewhat blocky.

### Animations

The UI should now feel more snappy as all delays caused by animations have been cut in half or sped up 2x depending
on how you want to look at it.

Also, the loading animations and how they were used have been changed so that the content jumps around less.
It still jumps around, but it does so less and that's good enough. 👍

## New and changed capabilities

As part of adding support for these new robots, Valetudo has been extended a bit

### AutoEmptyDockAutoEmptyIntervalControlCapability

This new core capability allows selecting the interval at which the dock should auto-empty the robot.

![image](https://github.com/Hypfer/Valetudo/assets/974410/25612d47-ff8c-4023-980f-bd7e8c3027e5)

Previously, this feature was already available as a quirk on some robots.

### CarpetSensorModeControlCapability

The CarpetSensorModeControlCapability has gained a new mode with this release: Detach

![image](https://github.com/Hypfer/Valetudo/assets/974410/c779345d-ace0-4388-b6c9-25185bc3e514)

This for now is only supported by the X40 and allows completely detaching the mop pads once it encounters a carpet.

### OperationModeControlCapability

You might've already spotted the fourth icon further up in the changelog but if not, here it is again:

![image](https://github.com/Hypfer/Valetudo/assets/974410/f1ac26d5-52dd-4c3c-8609-f42797b7b702)

The new operation mode "Vacuum then mop" does exactly what you'd think it does (if your robot's firmware supports it).
I have no clue why one would want that over doing both at the same time, but it is here now so have fun I guess.

Also, if you have a use case for it, feel free to leave a comment and let me know. I am genuinely curious.

## MQTT

This release comes with a **breaking change** for Home Assistant users.
Due to a schema change, you will have to run at least Home Assistant 2024.02 to use this version of Valetudo.

Given that it is June now, I think I've waited long enough to give everyone ample opportunity to update their setup to that version or later.

### More optionally exposable capabilities

As requested by a user, it is now possible to optionally expose the KeyLockCapability via MQTT as a dynamic defense against toddlers and cats.

![image](https://github.com/Hypfer/Valetudo/assets/974410/13d244ea-431a-429c-9065-489ad0cc1775)

Furthermore, it is now also possible to expose the ObstacleAvoidanceControlCapability to e.g. turn off obstacle avoidance via MQTT
if the robot is sent to a room where there might be problematic curtains or similar.

### MQTT fixes

This release includes an updated MQTT.js library, which should solve some weird and unexplainable reconnect issues in some setups.

Furthermore, a fix has been incorporated that should result in near-instant feedback when toggling stuff via MQTT.

## Misc

Thanks to a user living in a large non-square house, we've learned that 3irobotix maps can actually be larger than 800x800px.
We also learned that due to the map assumed to be square, in some places, width and height were swapped, leading to a weirdly sliced non-square map.

![image](https://github.com/Hypfer/Valetudo/assets/974410/2bc5c163-e93d-4a4d-9b04-682a1b8ea936)

This has now been fixed.

Lastly, an issue with timer pre_actions not being applied correctly should have been fixed by the use of `sleep()`.
The best way of fixing concurrency issues.

## On not burning out

I must say that working this much on improving Valetudo again has been a lot of fun.
I really enjoyed it, it gave me a lot of energy and I think it also produced great results.

What I **do not** enjoy however is that me gaining energy from maintaining the project is something so unusual, it's worth explicitly mentioning.

This is because most of the time I spend maintaining Valetudo, I spend fending off bullshit/abuse and having weekly internet fights over defending my personal boundaries.

- Be it nonsense discussions fueled by emotions/entitlement/hidden agenda.
- Be it callout posts on social media trying to frame me as some kind of bad guy for the mere act of saying "no".
- Be it content creators or other commercial entities trying to non-consensually monetize the project for their own benefit.

This comes on top of the daily grind consisting of but not being limited to:

- People not reading instructions
- People not reading the docs
- People reading the docs but then asking if the docs are really correct on the matter, because they did not like what they read
- People not doing their own homework
- People throwing the mental load they're responsible to carry at random people on the internet
- People asking questions they know the answer to but don't like and thus wanting to hear it again
- People trying to make their completely unreasonable emotions my problem
- People asking if unsupported robots are supported
- People asking when a new robot will be supported
- People not putting in the necessary effort required to practice this hobby

That said, I don't think that there is much point in whining about it, as that is just how it is if you run a project like this one.
The only thing you can do is improve your processes and learn how to better deal with it.

There is point though in documenting it and, most importantly, making it **visible**, which is why I did that just now.
And now that I did it, I think I'm good.

Thanks for reading!
And thanks a lot to the **vast majority** of the userbase that isn't like that. ❤

## The usual

If you want to see Valetudo on more robots and/or like this release, you might want to consider donating:

[https://github.com/sponsors/Hypfer](https://github.com/sponsors/Hypfer)

[https://builder.dontvacuum.me/donations.txt](https://builder.dontvacuum.me/donations.txt)

### Breaking Changes

- **mqtt**: Remove deprecated schema property from HA vacuum entity autodiscovery data [`ef92967`](http://github.com/Hypfer/Valetudo/commit/ef92967f7566bbe611fcfeb79e60f3f0f083d5ff)

### Features

- **ui**: HiDPI support for the map renderer [`86555ae`](http://github.com/Hypfer/Valetudo/commit/86555ae5757773bc580a6e550cd203cc82e5ba46)
- **ui**: Better feedback for the dock controls [`7a318cf`](http://github.com/Hypfer/Valetudo/commit/7a318cfb70bb1d8119e209dd39ff9fbb54ce3d7d)
- **mqtt**: Optionally expose ObstacleAvoidanceControlCapability [`9866d94`](http://github.com/Hypfer/Valetudo/commit/9866d9498763e14c5452c86cb0f8c772d1b56800)
- **mqtt**: Optionally expose KeyLockCapability [`572e3ce`](http://github.com/Hypfer/Valetudo/commit/572e3ce329ab6b8d92f9dfc9bb072ca3b8cbd8c6)
- **vendor.dreame**: R2416 support [`d87dca8`](http://github.com/Hypfer/Valetudo/commit/d87dca8028422528664a7f9aa6138a35ea2e61c0)
- OperationModeControlCapability: Vacuum then Mop Mode [`1937de9`](http://github.com/Hypfer/Valetudo/commit/1937de9b0fcd33d6cdda491a9b8d6cdd4fbc0701)
- CarpetSensorModeControlCapability: Detach Mode [`6c3dd7e`](http://github.com/Hypfer/Valetudo/commit/6c3dd7eae49e32f195ebe9395c40e3c86dd299c2)
- **ui**: Loading animation cleanup [`d8714dd`](http://github.com/Hypfer/Valetudo/commit/d8714ddeb97d20b59b2ccb56be72ed927d5f8c6d)
- **ui**: Refactor, redesign and rearrange controls a bit [`138f0d9`](http://github.com/Hypfer/Valetudo/commit/138f0d911c4d07a0839ad2a770c7951280a4c372)
- **ui**: Change desktop UI breakpoint from min 600px to min 750px width [`97818ac`](http://github.com/Hypfer/Valetudo/commit/97818acc183db36e5a3d87e77fee403fc7d35adc)
- **ui**: Custom icons for Fan speeds, Water grades and Operation modes [`03ff290`](http://github.com/Hypfer/Valetudo/commit/03ff2909a925235a4a7a972b931deb04b1983aba)
- AutoEmptyDockAutoEmptyIntervalControlCapability [`6465975`](http://github.com/Hypfer/Valetudo/commit/6465975d8edfdcba389593da629ef6bf6acfd81e)
- **vendor.dreame**: L10S Pro Ultra Heat support [`99e13da`](http://github.com/Hypfer/Valetudo/commit/99e13da0d5e49afda2dd0983df16f9314a3bec29)
- **vendor.roborock**: dryer module support for S8 Pro and G10s (#2037) [`37f386a`](http://github.com/Hypfer/Valetudo/commit/37f386a509f30230da4ffd0a9afa33aa06f96bea)
- Allow advanced users to customize their system ID [`a9868f7`](http://github.com/Hypfer/Valetudo/commit/a9868f720a4ad8979dbcd27af0b0c3f72a6c8a65)
- **ui**: Speed up all animations by 2x [`8f26ae7`](http://github.com/Hypfer/Valetudo/commit/8f26ae712b541e8e8b73bdc54260ca3a25b8415e)

### Fixes

- **ui**: Extending HTML elements is not supported by safari [`4f72632`](http://github.com/Hypfer/Valetudo/commit/4f72632e195c13f6110743461c776a92b5e3197a)
- **vendor.dreame**: Reduce repoll delay right after startup to reduce the time to first map on some dreames [`a7b2cc4`](http://github.com/Hypfer/Valetudo/commit/a7b2cc430c02e152636523529eb569d4d353d8c3)
- **ui**: Fix some hiDPI related issues [`13f7db3`](http://github.com/Hypfer/Valetudo/commit/13f7db381a5d09e0176c0e0b841d732454754000)
- **ui**: Draw stroke before fill for map structures that have text to fix rendering on firefox mobile [`7d9255b`](http://github.com/Hypfer/Valetudo/commit/7d9255b7f3c83aa5c9263e48ac4bccae8fa2e1be)
- **vendor.dreame**: Fix detergent property value 3 mapping [`09e2083`](http://github.com/Hypfer/Valetudo/commit/09e20834beb4b2eaa900c3015fe255158a196e77)
- **ui**: Don't display preset value until it is actually committed [`33615ec`](http://github.com/Hypfer/Valetudo/commit/33615ec190612ba4f77b8354932a7a2648d5d5cc)
- **vendor.dreame**: Warn on unknown preset value [`77be950`](http://github.com/Hypfer/Valetudo/commit/77be950a2c475bb95e9778da3e5d651d847a7024)
- **mqtt**: Always refresh gettable handles after setting [`6ca4eb9`](http://github.com/Hypfer/Valetudo/commit/6ca4eb940d8a59f4a4c2cd6f6f9a25d9b0bec453)
- **MockRobot**: Remove turbo water grade [`92b351a`](http://github.com/Hypfer/Valetudo/commit/92b351a1d91191d444d3a2667f01f7af423e5b41)
- **timers**: Wait a few seconds after executing pre_actions to give the firmware some time to report back its state [`f0101b7`](http://github.com/Hypfer/Valetudo/commit/f0101b7aabe9d542cb4090460607131d3091d95a)
- Don't crash the process if we're not able to persist the config [`cc0cd41`](http://github.com/Hypfer/Valetudo/commit/cc0cd41852d6d064cf18188bc8bdd5172426b602)
- **ui**: Fix missing controls card content indentation [`31519e1`](http://github.com/Hypfer/Valetudo/commit/31519e164c854e691b404b7478ecadf62c5c03d1)
- **mqtt**: Bump to mqttjs v5.5.5 to fix some keep-alive related issues [`9dab16c`](http://github.com/Hypfer/Valetudo/commit/9dab16c50ef1922e6888f2902f1dbdde22792573)
- **vendor.roborock**: The S6 does not support the MapSnapshotCapability [`7362514`](http://github.com/Hypfer/Valetudo/commit/7362514484faa4de86228ce3bff371fff80b9a6c)
- **ui**: Prevent the creation of virtual walls with a length of 0 [`b3417d2`](http://github.com/Hypfer/Valetudo/commit/b3417d2d0074f95387dfd9c1d28147641b5f6e6d)
- **vendor.dreame**: GoTo support has been removed from later X10plus firmwares [`b1f58ab`](http://github.com/Hypfer/Valetudo/commit/b1f58abfffa611e614af09675a8bbe71e8a30ce6)
- **vendor.3ir**: Fix parsing of maps with different dimensions than 800x800px [`648b07e`](http://github.com/Hypfer/Valetudo/commit/648b07e8434477aeb2ba9c6dfbd27f003505c76c)
- Fix ConfigPath env variable [`e7d2bc6`](http://github.com/Hypfer/Valetudo/commit/e7d2bc60d489403c5cba310dbe914918b672d2e1)
- **vendor.dreame**: Document error code 117 (#2020) [`4af21c7`](http://github.com/Hypfer/Valetudo/commit/4af21c7596b69a50a1ef0ef63e15858030836751)
- **vendor.dreame**: Remove mop drying time quirk from the W10 Pro as it is not implemented by the firmware [`28c91b6`](http://github.com/Hypfer/Valetudo/commit/28c91b65ad7af7a77fad32b84a6f6a75865446ba)

### Chores

- **release**: 2024.06.0 [`99cbf90`](http://github.com/Hypfer/Valetudo/commit/99cbf90ef19204908f34acc373ee5c165847d66b)
- Misc cleanup [`0a036ef`](http://github.com/Hypfer/Valetudo/commit/0a036eff5366ded02787de4f740350fc75cace5a)
- More random github action cron schedules [`4076f6a`](http://github.com/Hypfer/Valetudo/commit/4076f6aa7e5f708de54f99ea228059ced81bb512)
- **build**: Bump all github actions to node 20 [`17dadc7`](http://github.com/Hypfer/Valetudo/commit/17dadc7a606a18f90541827c97707d6e70a9a322)
- **ui**: Update TLS wording (#2012) [`97296b6`](http://github.com/Hypfer/Valetudo/commit/97296b6fdab3f37eef639234344789fe7af31bac)
- Discussion template [`fe6e6cc`](http://github.com/Hypfer/Valetudo/commit/fe6e6cc38a4ff1d7101e97b71aa18e01cfabff6e)
- Update workflow [`7b6f506`](http://github.com/Hypfer/Valetudo/commit/7b6f506bce15486c51a6d5f4b9190dea8b55f929)

<hr /><em>This discussion was created from the release <a href='https://github.com/Hypfer/Valetudo/releases/tag/2024.06.2'>Valetudo 2024.06.2</a>.</em>

**Comment by Pr0mises** [comment]
Just letting you know you rock. I still regret not meeting you at CCC

**Comment by Raphty** [comment]
I also wanted to say thank you for this Project, I know how people sometimes think that just because something is Open Source devs have to  respond to issues the way they think is OK.

You are doing great work here, and I appreciate it a lot! Thank you.

**Comment by aradove** [comment]
Hi!

Great work! For the newly added models, it hot water while mopping supported?

I see in the commit where l10s ultra heat was added
```
case DreameQuirkFactory.KNOWN_QUIRKS.MOP_DOCK_WATER_HEATER:
                return new Quirk({
                    id: id,
                    title: "Mop Dock Water Heater",
                    description: "When enabled, the dock will heat the water used to rinse the mop pads.",
```
Does it mean that it is hot water loaded in the robot before it cleans, or when it is docked with the station?

Sorry if it is a stupid question. I tried searching github issues, read this release statement, searching the website but I did not fully understand if it was supported. 

I was either going to buy the L10s Pro Ultra heat or the  L10s Pro Ultra this week :) 

Thank you for working on such an awesome project!

Edit: from product view it is only heating the mops when cleaning them in station. I missunderstood it!

**Reply by aradove** [reply]
I managed to join the telegram group and saw that it only means that it cleans the mops in the station with it. 

My misstake, I did not understand it properly.

Then I know, and both looks like solid alternatives! Sorry for disturbing!

**Comment by maddin8** [comment]
About the vacuuming, then mopping instead of doing both at the same time: My L10s Ultra, which is not running Valetudo yet but possibly will in the future, does scheduled vacuuming and then separate vacuuming and mopping runs to collect the largest parts of dust and dirt before it wets the floor and makes some of it stick.

There's sometimes dirt that gets wetened by the mops and them sticks to the ground so that the robot can't pick it up as good. 
That's why I like to have it do these tasks separately.
